### PR TITLE
Portable stories: Use internal Storybook renderers mechanism

### LINKED ISSUE #28685: Use internal Storybook renderer logic for portable stories in 8.3 (as a default), as it will solve hard to debug race conditions and allows us to run the vitest test runner for all renderers [kasper]
Portable stories was originally more of a way to get the resolved component of a story, and let users render that component themselves, e.g. with RTL (React Testing Library).

In 8.2 we now kind of mimic the whole story run in portable stories, but still leave the rendering up to the user. We can make the rendering part of Storybook also portable, but we just haven’t done that yet.

One big difference between our internal rendering logic and RTL is that our internal logic is async and waits on more stuff before starting the play function. This can give subtle but frustrating differences in how the story will run. Think about that `getByRole` will resolve in Storybook but fail in portable stories.

Note that this would make portable stories more reliable, ensure that we can play stories of all our renderers, and eliminate the need for the user to configure portable stories with testing library.

### PR DESCRIPTION
**Pull request by kasperpeulen**
Closes #28685

#### The changes in this PR are covered in the following automated tests:
- [ ] stories
- [ ] unit tests
- [ ] integration tests
- [ ] end-to-end tests

#### Manual testing

_This section is mandatory for all contributions. If you believe no manual test is necessary, please state so explicitly. Thanks!_

### Documentation

- [ ] Add or update documentation reflecting your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

- [ ] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli/src/sandbox-templates.ts`
- [ ] Make sure this PR contains **one** of the labels below:
   <details>
     <summary>Available labels</summary>

     - `bug`: Internal changes that fixes incorrect behavior.
     - `maintenance`: User-facing maintenance tasks.
     - `dependencies`: Upgrading (sometimes downgrading) dependencies.
     - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
     - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
     - `documentation`: Documentation **only** changes. Will not show up in release changelog.
     - `feature request`: Introducing a new feature.
     - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
     - `other`: Changes that don't fit in the above categories.
   
   </details>

### 🦋 Canary release

This PR does not have a canary release associated. You can request a canary release of this pull request by mentioning the `@storybookjs/core` team here.

_core team members can create a canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=<PR_NUMBER>`_

| name | before | after | diff | z | % |
| ---- | ------ | ----- | ---- | - | - |
| createSize |  0 B | 0 B | 0 B | - | - |
| generateSize |  76.3 MB | 76.3 MB | 0 B | -0.03 | 0% |
| initSize |  198 MB | 198 MB | -16.1 kB | 0.53 | 0% |
| diffSize |  122 MB | 122 MB | -16.1 kB | 0.78 | 0% |
| buildSize |  7.62 MB | 7.61 MB | -11 kB | -0.05 | -0.1% |
| buildSbAddonsSize |  1.64 MB | 1.63 MB | -6.91 kB | -0.83 | -0.4% |
| buildSbCommonSize |  195 kB | 195 kB | 0 B | - | 0% |
| buildSbManagerSize |  2.31 MB | 2.31 MB | -461 B | 0.47 | 0% |
| buildSbPreviewSize |  351 kB | 349 kB | -2.04 kB | **-1.4** | -0.6% |
| buildStaticSize |  0 B | 0 B | 0 B | - | - |
| buildPrebuildSize |  4.5 MB | 4.49 MB | -9.41 kB | -0.16 | -0.2% |
| buildPreviewSize |  3.13 MB | 3.13 MB | -1.6 kB | 0.36 | -0.1% |
| testBuildSize |  0 B | 0 B | 0 B | - | - |
| testBuildSbAddonsSize |  0 B | 0 B | 0 B | - | - |
| testBuildSbCommonSize |  0 B | 0 B | 0 B | - | - |
| testBuildSbManagerSize |  0 B | 0 B | 0 B | - | - |
| testBuildSbPreviewSize |  0 B | 0 B | 0 B | - | - |
| testBuildStaticSize |  0 B | 0 B | 0 B | - | - |
| testBuildPrebuildSize |  0 B | 0 B | 0 B | - | - |
| testBuildPreviewSize |  0 B | 0 B | 0 B | - | - |

| name | before | after | diff | z | % |
| ---- | ------ | ----- | ---- | - | - |
| createTime |  23s | 24s | 1s | **2.15** | 4.2% |
| generateTime |  21.9s | 22.5s | 606ms | 0.26 | 2.7% |
| initTime |  24.6s | 23.4s | -1s -176ms | 0.61 | -5% |
| buildTime |  15.3s | 15.8s | 413ms | **1.8** | 2.6% |
| testBuildTime |  0ms | 0ms | 0ms | - | - |
| devPreviewResponsive |  8.6s | 8.6s | 28ms | -0.29 | 0.3% |
| devManagerResponsive |  5.5s | 5.7s | 195ms | -0.33 | 3.4% |
| devManagerHeaderVisible |  787ms | 834ms | 47ms | -0.14 | 5.6% |
| devManagerIndexVisible |  818ms | 867ms | 49ms | -0.16 | 5.7% |
| devStoryVisibleUncached |  1.3s | 1.3s | 3ms | -0.66 | 0.2% |
| devStoryVisible |  839ms | 894ms | 55ms | -0.13 | 6.2% |
| devAutodocsVisible |  750ms | 805ms | 55ms | 0.16 | 6.8% |
| devMDXVisible |  720ms | 817ms | 97ms | -0.13 | 11.9% |
| buildManagerHeaderVisible |  755ms | 711ms | -44ms | **-1.76** | 🔰-6.2% |
| buildManagerIndexVisible |  758ms | 716ms | -42ms | **-1.64** | 🔰-5.9% |
| buildStoryVisible |  804ms | 758ms | -46ms | **-1.71** | 🔰-6.1% |
| buildAutodocsVisible |  709ms | 658ms | -51ms | -1.19 | -7.8% |
| buildMDXVisible |  697ms | 655ms | -42ms | -0.74 | -6.4% |

## Greptile Summary

This pull request introduces an internal `renderToCanvas` method as a fallback when `testingLibraryRender` is not specified, along with several related changes.

- **`code/core/src/preview-api/modules/store/csf/portable-stories.ts`**: Removed fallback logic for `renderToCanvas` and introduced a new `run` function for story execution.
- **`code/core/src/preview-errors.ts`**: Removed `TestingLibraryMustBeConfiguredError` class to simplify error handling.
- **`code/core/src/types/modules/composedStory.ts`**: Added optional `play` and mandatory `run` functions to `ComposedStoryFn` type.
- **`code/lib/react-dom-shim/src/preventActChecks.tsx`**: Added `preventActChecks` function to temporarily disable `IS_REACT_ACT_ENVIRONMENT` flag.
- **`code/renderers/react/src/portable-stories.tsx`**: Introduced conditional check to use internal `renderToCanvas` method if `testingLibraryRender` is not specified.

### CODE REVIEW

**Review by yannbf — APPROVED** [review]

File: code/renderers/svelte/src/__test__/composeStories/__snapshots__/portable-stories.test.ts.snap [hunk 1]
```diff
@@ -152,6 +173,9 @@ exports[`Renders NewStory story 1`] = `
     </div>
     <!---->
     <!---->
+    <!---->
+    <!---->
+    <!---->
```
[the remark this answers is not in the document]
Comment (kasperpeulen) [reply code/renderers/svelte/src/__test__/composeStories/__snapshots__/portable-stories.test.ts.snap:178]: Yes, those are coming from PreviewRender that was first not used in the testing library setup

File: code/lib/react-dom-shim/src/react-18.tsx [hunk 2]
```diff
@@ -21,20 +23,33 @@ const WithCallback: FC<{ callback: () => void; children: ReactElement }> = ({
   return children;
 };
 
+// pony-fill
+if (typeof Promise.withResolvers === 'undefined') {
```
[the remark this answers is not in the document]
Comment (kasperpeulen) [reply code/lib/react-dom-shim/src/react-18.tsx:27]: Will do, it is a polyfill for `Promise.withResolvers`

File: code/renderers/react/src/__test__/Button.stories.tsx [hunk 3]
```diff
@@ -87,6 +87,34 @@ export const CSF3ButtonWithRender: CSF3Story<ButtonProps> = {
...
+        <input data-testid="input" />
+        <br />
+        <button onClick={() => setClicked(!isClicked)}>
+          I am {isClicked ? 'clicked' : 'not clicked'}
+        </button>
+      </>
+    );
+  },
+  play: async ({ canvasElement, step }) => {
+    console.log('start of play function');
```
[the remark this answers is not in the document]
Comment (kasperpeulen) [reply code/renderers/react/src/__test__/Button.stories.tsx:104]: I don't know, I think you added them, I copied it from the test sandbox

File: test-storybooks/portable-stories-kitchen-sink/nextjs/stories/__snapshots__/portable-stories.test.tsx.snap [hunk 4]
```diff
@@ -8,6 +8,9 @@ exports[`renders dynamicImportStories stories renders Default 1`] = `
     >
       Global Decorator
       <br />
+      <div>
+        I am a dynamically loaded component
```
[the remark this answers is not in the document]
Comment (kasperpeulen) [reply test-storybooks/portable-stories-kitchen-sink/nextjs/stories/__snapshots__/portable-stories.test.tsx.snap:12]: Yes, this is one of the differences between how we render async, and testing-library async. And an example of mismatches that we prevent now!

File: code/renderers/react/src/__test__/Button.stories.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (yannbf) [reply code/renderers/react/src/__test__/Button.stories.tsx:104]: feel free to remove them!

File: code/renderers/react/src/__test__/portable-stories.test.tsx [hunk 5]
```diff
@@ -14,7 +14,7 @@ import { setProjectAnnotations, composeStories, composeStory } from '..';
 import type { Button } from './Button';
 import * as stories from './Button.stories';
 
-setProjectAnnotations([{ testingLibraryRender: render }]);
+setProjectAnnotations([]);
```
[the remark this answers is not in the document]
Comment (kasperpeulen) [reply code/renderers/react/src/__test__/portable-stories.test.tsx:17]: setProjectAnnotations still register the react annoations. So I need to call it. And TS accepts only accepts an empty array, we probably should change that though?

**Pull request merged**